backend/database/Search.py

```diff
@@ -1,38 +1,121 @@
+                              " OR devices_tags.tag_id = " + token[1:] + \
+                              " OR rooms_tags.tag_id = " + token[1:] + \
+                              " OR buildings_tags.tag_id = " + token[1:] + ")"
+            elif re.match("#\\d+", token) and search_type == 'device':
+                sql_string += " (devices_tags.tag_id = " + token[1:] + \
+                              " OR rooms_tags.tag_id = " + token[1:] + \
+                              " OR buildings_tags.tag_id = " + token[1:] + ")"
+            elif re.match("#\\d+", token) and search_type == 'room':
+                sql_string += " (rooms_tags.tag_id = " + token[1:] + \
+                              " OR buildings_tags.tag_id = " + token[1:] + ")"
+            elif re.match("#\\d+", token) and search_type == 'building':
+                sql_string += " buildings_tags.tag_id = " + token[1:]
+            elif re.match("\\$\\d+", token):
+                sql_string += " rooms.room_id = " + token[1:]
+            elif re.match("%\\d+", token):
+                sql_string += " devices.device_id = " + token[1:]
+            elif re.match("\\*\\d+", token):
+                sql_string += " points.point_id = " + token[1:]
+            elif re.match(":floor ([<>=]|!=|<=|>=) (\\d+)", token):
+                matches = re.match(":floor ([<>=]|!=|<=|>=) (\\d+)", token).groups()
+                sql_string += " rooms.floor " + matches[0] + " " + matches[1]
+            elif re.match(":type (\\d+)", token):
+                matches = re.match(":type (\\d+)", token).groups()
+                sql_string += " points.value_type_id = " + matches[0]
+            elif re.match(":unit (\\d+)", token):
+                matches = re.match(":unit (\\d+)", token).groups()
+                sql_string += " value_units.value_unit_id = " + matches[0]
+            elif re.match(":measurement (\'\\w+\')", token):
+                matches = re.match(":measurement (\'\\w+\')", token).groups()
+                sql_string += " value_units.measurement = " + matches[0]
+            elif re.match("and", token):
+                sql_string += " AND"
+            elif re.match("or", token):
+                sql_string += " OR"
+            elif re.match("not", token):
+                sql_string += " NOT"
+            elif re.match("\(", token):
+                sql_string += "("
+            elif re.match("\)", token):
+                sql_string += ")"
 
         return sql_string
+
+    @staticmethod
+    def points(source_string):
+        """Uses the given search string to generate a SQL WHERE-clause to be used in a query for
+        points.
+        """
+        # @ -> buildings, # -> tags, $ -> rooms, % -> devices, * -> points
+        regex = re.compile("^([@#$%*]\\d+|and|or|not|:(floor|type|unit|measurement) (([<>=]|!=|<=|>=)? ?(\\d+)|(\'\\w+\'))|\\(|\\)|\\s+)+$")
+
+        if regex.match(source_string) is None:
+            raise Exception("Invalid source string for points.")
```

Comment:
Will this catch all invalid source strings, or are there ones that will match the regex but still throw an error in the DB? In either case, can we throw this as a HTTP 400 error rather than letting it crash as a 500 so the user knows their request was at fault?

Reply:
We created a custom exception for this and are now aborting as a 400 if it's raised. Good idea!